### ISSUE TITLE
More precise type inference for records

### ISSUE THREAD

**Issue by Kukovec** [informalsystems/apalache#401]
We should rethink type inference for sets of records of different types.
One possible direction would be to introduce something similar to variant records (https://en.wikibooks.org/wiki/Ada_Programming/Types/record#Variant_record). We could then treat expressions of the form `r.type = X` as casts, possibly accompanied by a change in the internal representation.

**Comment by konnov** [comment]
Here is a piece of text from https://apalache.informal.systems/docs/adr/002adr-types.html for the context:

> Note: For the above example of a set of records, we are considering to introduce union types. So the type of the set
> 
> 
> {[type |-> "1a", bal |-> 1],
>  [type |-> "2a", bal |-> 2, val |-> 3]}
> would be something like:
> 
> 
> Set([type |-> "1a"], bal |-> 1] + [type |-> "2a", bal |-> 2, val |-> 3])
> The value of the field type would serve as a type tag. However, we would have to fix a set of patterns that turn a union type into a precise record type. One such pattern is a set comprehension, e.g., { r \in S: r.type = "1a" }. If you have suggestions on this line of thought, please let us know.

**Comment by konnov** [comment]
It looks like at least three users ran into hard-to-debug issues that would not appear if we had a precise record type checker.

**Comment by konnov** [comment]
This seems to be a usability issue after all

**Comment by konnov** [comment]
Let's figure patterns first and write an ADR in addition to https://apalache.informal.systems/docs/adr/002adr-types.html

**Comment by shonfeder** [comment]
Might we also consider using https://en.wikipedia.org/wiki/Row_polymorphism ? This would just enable the (imo) most natural sub-typing relationship on records, where a record `r1 <: r2` iff `r2` has all the `fieldName : type`s in `r1` (but possibly others as well). I think that would cover  the case you gave, without having to introduce a special meaning for a particular key.

**Comment by konnov** [comment]
I am not sure about row polymorphism, but I guess it is different. What we need are records that have different combinations of fields and are distinguished by a tag field. That's the pattern introduced by Lamport, which would be very hard to replace with something else.

**Comment by shonfeder** [comment]
What would be the difficulty in replacing this with a more principled approach?

**Comment by konnov** [comment]
One thing is to add a few additional tests where needed, another one is to completely rewrite a spec. Here is a canonical example of how records are mixed in a set: https://github.com/tlaplus/Examples/blob/master/specifications/Paxos/Paxos.tla

As far as I understand row polymorphism is that one can shrink a record type. The way Leslie uses records (and thus all people who are reading his book and copying his specs) is using them as unions in C.

**Comment by konnov** [comment]
If we look at this set from Paxos `{[type |-> "1a", bal |-> 1], [type |-> "2a", bal |-> 2, val |-> 3]}`, then the way records are used here is similar to algebraic data types. For instance, in Scala we would define two case classes:

```scala
case class Msg1a(bal: Int) extends PaxosMessage {}

case class Msg2a(bal: Int, val: Int) extends PaxosMessage {}
```

If we agree that the field `type` specifies the constructor of such a type, we could treat the two records in Paxos as two objects that are constructed by using two different constructors. So we could introduce a new kind of a type definition for those variant/inductive/algebraic types in type annotations:

```
@typeDef: MSG =
          [type -> "1a", bal: Int]
        | [type -> "2a", bal: Int, val: Int];
```

So we can construct records of two types:

 - a record `r1` of type `[type: Str, bal: Int]` that has `r1.type = "1a"`.
 - a record `r2` of type `[type: Str, bal: Int, val: Int]` that has `r2.type = "2a"`.

We can discuss the concrete syntax here, but that is basically an ADT. The type of the above set will be `Set(MSG)`.

We can go even further and sneak in recursive/inductive data structures:

```
@typeDef: TREE =
          type = "Leaf" -> [type: Str, val: Int]
        | type = "Node" -> [type: Str, left: TREE, right: TREE];
```

**Comment by lemmy** [comment]
FWIW: An example of a spec with the message types `{[type |-> "tok", q |-> 1, color |-> "black"], [type |-> "pl", src |-> "addr"]}` , and where the refinement mapping drops a (record) field from the second msg type: https://github.com/tlaplus/Examples/blob/22e6b2daae5e0173ac856e8eab5c01abd90b39d4/specifications/ewd998/EWD998ChanID.tla#L183-L189

**Comment by shonfeder** [comment]
Ah, I see. Thanks for the explanation @konnov .  I think I misunderstood the need for backwards compatibility here, and so how much leeway we had when considering a more principled type discipline around records.

**Comment by konnov** [comment]
Since there is no pattern matching in TLA+, we have to figure out how to deconstruct those types. The obvious patterns are like: `IF m.type = "1a" THEN … ELSE …` and `{ m \in Msgs: m.type = "1a" }`. These two patterns are consistent with the record idioms that we see in the examples. We could also use `CASE` to do pattern matching over union types. We have to figure out these patterns and see, whether they are comprehensive.

**Comment by konnov** [comment]
Looking at the [allocator example](https://github.com/tlaplus/Examples/blob/616c4c2e00dd7084c623d1dcc83b140279652fb4/specifications/allocator/AllocatorImplementation.tla#L90-L93), we would probably have to introduce a pattern for converting a generic message to a special message, by looking at `m.type`. @Kukovec this is what you probably meant by a casting operator, right?

I am trying to come up with a reasonable solution and that is not easy. This is the best one I have so far:

```tla
RAlloc(m) ==
  /\ m \in network
  /\ \E ma \in {m}:
      /\ ma.type = "allocate"
      /\ holding' = [holding EXCEPT ![ma.clt] = @ \cup ma.rsrc]
      /\ requests' = [requests EXCEPT ![ma.clt] = @ \ ma.rsrc]
```

**Comment by Kukovec** [comment]
You could try set filtering
```
\E ma \in {msg \in {m} : msg.type = "allocate" }: ...
```
Also, you could implement this to be the body of the "casting" operator:
```
Cast(rec, fld) = CHOOSE cast \in { r \in {rec} : r.type = fld } : TRUE
```

**Comment by konnov** [comment]
This option looks interesting, though it is heavy on syntax:

```tla
\E ma \in {msg \in {m} : msg.type = "allocate" }: ...
```

I am not sure how to use `Cast(rec, fld)`. Can you try to use in that above example?

The problem is that we both have to check the type and introduce a new name for a matching message. There is a small number of operators in TLA+ that allow such a trick.

**Comment by Kukovec** [comment]
In the above, you could do 
```
LET mCast == Cast(m, "allocate")
IN 
  /\ mCast \in network
  /\ holding' = [holding EXCEPT ![mCast.clt] = @ \cup mCast.rsrc]
  /\ requests' = [requests EXCEPT ![mCast.clt] = @ \ mCast.rsrc]
  
```

**Comment by Kukovec** [comment]
It shouldn't be impossible to LET-wrap as a preprocessing step.

**Comment by konnov** [comment]
Ah, I was also thinking about a let expression. I guess, it is syntactically concise. Though if you just cast, we need an additional test for the actual type. Cast would just shrink the type. By the way, `CHOOSE` is free to return whatever on the empty set.

**Comment by Kukovec** [comment]
Yes, but what else do you do on a bad cast?

**Comment by konnov** [comment]
`RAlloc(m)` is a predicate, so it should produce `FALSE`, when `m.type` is different from `"allocate"`.

**Comment by Kukovec** [comment]
Sure, then you can jut wrap with an ITE:
```
IF m.type = "allocate"
THEN (cast)
ELSE FALSE
```

**Comment by konnov** [comment]
Sure, you can. But the more baroque this pattern is, the more the chances to get it wrong.

**Comment by konnov** [comment]
How about the following syntactic form:

```tla
RAlloc(m) ==
  /\ m \in network
  /\ m.type = "allocate" => \E ma \in {m}:
        /\ holding' = [holding EXCEPT ![ma.clt] = @ \cup ma.rsrc]
        /\ requests' = [requests EXCEPT ![ma.clt] = @ \ ma.rsrc]
```

Although it looks a bit unusual, it is somewhat similar to pattern matching in Scala :-)

**Comment by konnov** [comment]
Of course, we can wrap the above example with a custom operator. I just want to see how we could avoid that, as a custom operator immediately introduces a module dependency on Apalache.

**Comment by konnov** [comment]
I have opened a discussion #789 on that, as the discussion in this issue is deeply buried. I have written five code patterns and now see what has worried me. Three of these patterns are not closed under negation.

**Comment by konnov** [comment]
I am closing this meta-issue. This work is finished. There are a few follow-up issues: #1926, #1943, #1944, #1945. All of them are tracked in the milestone: https://github.com/informalsystems/apalache/milestone/60

**Comment by shonfeder** [comment]
Hurrah! Awesome work here :)
